### PR TITLE
Add persist API

### PR DESCRIPTION
**Pull request by TBonnin**
First commit implementing the persist API.
The goal is to remove the database dependency for the runner. Writing customers data (batchSave/.../activityLogs/lastSyncDate) will go through this API
All of it behind a feature flag

Next PR: take care of removing db dependency for proxy.

## Issue ticket number and link
https://linear.app/nango/issue/NAN-270/persist-api

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [x] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is: will do in follow up PR
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/persist/lib/server.unit.test.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                    query.response({ id: 1 });
+                },
+                () => {
+                    query.response([]);
+                }
+            ];
+            steps[step - 1]?.();
+        };
+    }
+}
```
Comment (TBonnin) [inline packages/persist/lib/server.unit.test.ts:196]: ideally I would like to have a real in-memory database that can be used in tests. But knex in-memory solution is to use sqlite which doesn't support our schema.
So for now I am just mocking the db queries

File: packages/persist/lib/server.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+server.put('/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records', validateRecordsRequest, persistController.updateRecords);
+
+// error handling middleware
+server.use((err: Error, _req: Request, res: Response, next: NextFunction) => {
+    if (err) {
+        res.status(500).json({ error: err.message });
+    } else {
+        next();
+    }
+});
```
Comment (TBonnin) [inline packages/persist/lib/server.ts:87]: do we want some sort of authorization mechanism to make sure requests are legit and actually made from a runner?

File: packages/persist/lib/server.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+server.post(
+    '/environment/:environmentId/log',
+    validateRequest({
+        params: z.object({
+            environmentId: z.string().transform(Number).pipe(z.number().int().positive()) as unknown as z.ZodNumber
+        }),
+        body: z.object({
+            activityLogId: z.number(),
+            level: z.enum(['info', 'debug', 'error', 'warn', 'http', 'verbose', 'silly']),
```
Comment (TBonnin) [inline packages/persist/lib/server.ts]: reuse the existing LogLevel enum

File: .github/workflows/deploy.yaml [hunk 4]
```diff
@@ -76,4 +77,20 @@ jobs:
...
+            - uses: docker/login-action@v3
+              with:
+                username: ${{ secrets.DOCKER_USERNAME }}
+                password: ${{ secrets.DOCKER_PASSWORD }}
+            - name: Push tag
+              run: |
+                docker buildx imagetools create nangohq/nango-persist:${{ github.sha }} --tag nangohq/nango-persist:${{ inputs.stage }}
+            - name: Deploy persist
+              run: |
+                SERVICE_ID=${{ fromJson('{ production: "srv-XXXXXXXXXXXXX", staging: "srv-cmsfiqqcn0vc73bhcod0" }')[inputs.stage] }}
```
Comment (khaliqgant) [inline .github/workflows/deploy.yaml]: We have this prod value now right?

File: packages/persist/lib/server.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import { validateRequest } from 'zod-express';
+import { z } from 'zod';
+import persistController from './controllers/persist.controller.js';
+import './tracer.js';
+import { logLevelValues } from '@nangohq/shared';
+
+export const server = express();
+server.use(express.json());
+
+// logging middleware
```
Comment (khaliqgant) [inline packages/persist/lib/server.ts]: can the non helpful comments be removed?

File: packages/persist/lib/server.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+server.use((req: Request, res: Response, next: NextFunction) => {
+    next();
+    console.log(`[Persist] ${req.method} ${req.path} ${res.statusCode}`);
+});
+
+// routes
```
Comment (khaliqgant) [inline packages/persist/lib/server.ts]: ^ remove

File: packages/persist/lib/server.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+});
+server.post('/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records', validateRecordsRequest, persistController.saveRecords);
+server.delete(
+    '/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records',
+    validateRecordsRequest,
+    persistController.deleteRecords
+);
+server.put('/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records', validateRecordsRequest, persistController.updateRecords);
+
+// error handling middleware
```
Comment (khaliqgant) [inline packages/persist/lib/server.ts]: ^ remove

File: packages/persist/lib/server.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/persist/lib/server.ts:87]: If the persist service is a private service then I don't think that is necessary yet.

**Review by khaliqgant — APPROVED** [review]
Looks good. Couple of non blocking comments and the note to add the persist server address in the github action

File: .github/workflows/deploy.yaml (hunk 4 above)
Comment (TBonnin) [reply .github/workflows/deploy.yaml]: yep. let me update it

File: packages/shared/lib/sdk/sync.ts [hunk 8]
```diff
@@ -13,8 +13,9 @@ import { Nango } from '@nangohq/node';
 import configService from '../services/config.service.js';
 import paginateService from '../services/paginate.service.js';
 import proxyService from '../services/proxy.service.js';
-
-type LogLevel = 'info' | 'debug' | 'error' | 'warn' | 'http' | 'verbose' | 'silly';
+import axios from 'axios';
+import { getPersistAPIUrl } from '../utils/utils.js';
+import type { LogLevel } from '../models/Activity.js';
```
Comment (khaliqgant) [inline packages/shared/lib/sdk/sync.ts:18]: @TBonnin FYI any type needs to not be imported which is why all the types are in this file. This causes this error as reported in [Slack](https://nangohq.slack.com/archives/C06GHQPLTAN/p1706859608331169)

**Pull request merged**